pennylane/measurements/counts.py

```diff
@@ -135,7 +140,125 @@ def circuit(x):
             )
         wires = Wires(wires)
 
+    # TODO: Remove this conditional branch when using `Counts.process` in devices
     if all_outcomes:
-        return MeasurementProcess(AllCounts, obs=op, wires=wires)
-
-    return MeasurementProcess(Counts, obs=op, wires=wires)
+        return _Counts(AllCounts, obs=op, wires=wires, all_outcomes=all_outcomes)
+
+    return _Counts(Counts, obs=op, wires=wires, all_outcomes=all_outcomes)
+
+
+# TODO: Make public when removing the ObservableReturnTypes enum
+class _Counts(SampleMeasurement):
+    """Measurement process that returns the samples of a given observable."""
+
+    def __init__(
+        self,
+        return_type: ObservableReturnTypes,
+        obs: Union[Observable, None] = None,
+        wires=None,
+        eigvals=None,
+        id=None,
+        all_outcomes=False,
+    ):
+        self.all_outcomes = all_outcomes
+        super().__init__(return_type, obs, wires, eigvals, id)
+
+    def process_samples(
+        self,
+        samples: Sequence[complex],
+        wire_order: Wires,
+        shot_range: Tuple[int] = None,
+        bin_size: int = None,
+    ):
+        samples = qml.sample(op=self.obs, wires=self._wires).process_samples(
+            samples, wire_order, shot_range, bin_size
+        )
+
+        if bin_size is None:
+            return self._samples_to_counts(samples)
+
+        if self.obs is None:
+            num_wires = qml.math.shape(samples)[0]
+            shape = (-1, bin_size, num_wires)
+        else:
+            num_wires = qml.math.shape(samples)[-1]
```

Comment:
This looks like it's unused

Reply:
Now that we have the `wire_order` argument I use that to compute `num_wires`

Reply:
We still need this conditional to define the shape though

Reply:
I was talking specifically about the `else` block and the unused `num_wires` variable, but it looks like you already changed it.